components/formats-api/src/loci/formats/MetadataList.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * #%L
+ * primary reader and writer APIs
+ * %%
+ * Copyright (C) 2018 Open Microscopy Environment:
+ *   - Board of Regents of the University of Wisconsin-Madison
+ *   - Glencoe Software, Inc.
+ *   - University of Dundee
+ * %%
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+
+package loci.formats;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.ArrayList;
+
+/**
+ * A container of metadata for FormatReader and other reader implementations.  The structure is designed to contain
+ * data for individual image series', including sub-resolutions.  This is implemented internally as a list of
+ * lists indexed by series and resolution index.  However, it may be used to store any data in a two-level
+ * structure.
+ *
+ * The container may be sized up front at the primary or at both levels.  Or either level may be appended to
+ * individually.  Both patterns are used by many different readers.
+ *
+ * This class was originally developed to store the CoreMetadata for image series and sub-resolutions.  However, it
+ * may be used to store any other data, for example reader-specific data including IFD mappings, extended
+ * metadata, etc.
+ *
+ * @param <T> The type of metadata to store.
+ */
+public class MetadataList<T> {
+  protected List<List<T>> data = new ArrayList<>();
```

Comment:
It wouldn't mean anything to add a third element right now.  The split is series/resolution, and this will be sufficiently future-proof for subresolutions in z.

If we want to get more complex, I think we would have to move to an entirely new interface.  This is all an artefact of being backward compatible with the old FormatReader API when adding subresolutions; ideally we would remove most of the coreIndex-related complexity entirely, but that's not currently an option.